crates/matrix-sdk-ui/tests/integration/sync_service.rs

```diff
@@ -202,3 +205,123 @@ async fn test_sync_service_state() -> anyhow::Result<()> {
+        assert_next_eq!(states, State::Running);
+
+        let next_state = timeout(states.next(), Duration::from_millis(500))
+            .await
+            .expect("We should have changed states by now")
+            .unwrap();
+
+        assert_eq!(next_state, State::Offline, "We should have entered the offline mode");
+    }
+
+    mock_server.mock_versions().ok().expect(1..).mount().await;
+
+    let next_state = timeout(states.next(), Duration::from_millis(500))
+        .await
+        .expect("We should changed the state")
+        .unwrap();
+
+    assert_eq!(next_state, State::Running, "We should have continued to sync");
+}
+
+#[async_test]
+async fn test_sync_service_offline_mode_stopping() {
+    let mock_server = MatrixMockServer::new().await;
+    let client = mock_server.client_builder().build().await;
+
+    let sync_service = SyncService::builder(client).with_offline_mode().build().await.unwrap();
+    let mut states = sync_service.state();
+
+    Mock::given(SlidingSyncMatcher)
+        .respond_with(ResponseTemplate::new(404))
+        .expect(1..)
+        .mount(mock_server.server())
+        .await;
+    mock_server.mock_versions().error500().mount().await;
+
+    sync_service.start().await.expect("We should be able to start the sync service");
+    assert_next_eq!(states, State::Running);
+
+    let next_state = timeout(states.next(), Duration::from_millis(500))
+        .await
+        .expect("We should have changed states by now")
+        .unwrap();
+
+    assert_eq!(next_state, State::Offline, "We should have entered the offline mode");
+
+    sync_service
+        .stop()
+        .await
+        .expect("We should be able to stop the sync service when it's in the offline mode");
+
+    let next_state = timeout(states.next(), Duration::from_millis(500))
+        .await
+        .expect("We should have changed the state by now")
+        .unwrap();
+
+    assert_eq!(next_state, State::Idle, "We should have entered the idle mode");
+}
+
+#[async_test]
+async fn test_sync_service_offline_mode_restarting() {
+    let mock_server = MatrixMockServer::new().await;
+    let client = mock_server.client_builder().build().await;
+
+    let sync_service = SyncService::builder(client).with_offline_mode().build().await.unwrap();
+    let mut states = sync_service.state();
+
+    Mock::given(SlidingSyncMatcher)
+        .respond_with(ResponseTemplate::new(404))
+        .mount(mock_server.server())
+        .await;
+    mock_server.mock_versions().error500().mount().await;
+
+    sync_service.start().await.expect("We should be able to start the sync service");
+    assert_next_eq!(states, State::Running);
+
+    let next_state = timeout(states.next(), Duration::from_millis(500))
+        .await
+        .expect("We should have changed states by now")
+        .unwrap();
+
+    assert_eq!(next_state, State::Offline, "We should have entered the offline mode");
+
+    sync_service
+        .start()
```

Comment:
`.restart()`?

Reply:
There is no `SyncService::restart()` method, the `SyncServiceInner` struct has such a method which is called by the `SyncService::start()` method if we're in the offline state.